src/basic/log.c

```diff
@@ -1157,15 +1157,36 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
         return 0;
 }
 
+static bool should_parse_proc_cmdline(void) {
+        const char *e;
+
+        /* PID1 always reads the kernel command line. */
+        if (getpid_cached() == 1)
+                return true;
+
+        /* If the process is directly executed through a unit, then reads the command line.
+         * If it is invoked by a script executed in a unit, then skip to read the command line. */
+        e = getenv("SYSTEMD_EXEC_PID");
+        if (e) {
+                pid_t p;
+
+                if (parse_pid(e, &p) >= 0)
+                        return getpid_cached() == p;
+                else
+                        log_warning("Failed to parse $SYSTEMD_EXEC_PID=%s. Ignoring.", e);
+        }
+
+        /* Assume that anything that has a controlling tty is user stuff. In that case, skip to read
+         * the command line. */
+        return get_ctty_devnr(0, NULL) < 0;
```

Comment:
If this check is necessary, that means that the issues that https://github.com/systemd/systemd/commit/41d1f469cf10f5f3e9cb4f4853ace9b0cfe5beae was trying to resolve are still present.
I think the logic here is still not strong enough: we cannot assume that $SYSTEMD_EXEC_PID will be propagated to all child processes. Thus, we should a) make sure to always set $SYSTEMD_EXEC_PID when appropriate, b) treat the *lack* of this variable as authoritative. So instead of the two checks, just something like this:

```C
 e = getenv("SYSTEMD_EXEC_PID");
 if (!e)
          return false;
 pid_t p;
 if (parse_pid(e, &p) < 0) {
         /* We know that systemd sets the variable correctly. Something else must have set it. */
         log_debug("Failed to parse \"SYSTEMD_EXEC_PID=%s\". Ignoring.", e);
         return false;
  }
 return getpid_cached() == p;
```

--

Also, it would be nice to be able to be able to opt-in, for testing and similar. So maybe add a special value:
```C
...
if (streq(e, "*"))
     return true;
...
```

Reply:
The reason I keep the condition `get_ctty_devnr() < 0` is for generators or other invocations by `execute_directories()`. Moreover, there are many other invocations by PID1, e.g. systemd-shutdown. I do not think it is a good idea we assume that `$SYSTEMD_EXEC_PID=` is set for all invoked process, as it may be easily forgotten when we introduce new `execve()`. Also, I cannot find any reason we need to ignore the kernel command line by CLI tools invoked in generators.

But there is a concern when a script including `systemctl` is invoked by `cron`...

Reply:
Yeah, but the `get_ctty_devnr(0, NULL) < 0` check is just unreliable. Nowadays many things are moving to executed in units under the user manager, and those command-line tools when invoked there will do the wrong thing. Getting all execve invocations right requires some work, but I think it shouldn't be so bad. As a general rule, we don't use execve to integrate functionality, but run everything from the same processes using libraries to compose code. So the number of execve's that we do is limited.

Reply:
@filbranden comments? IIUC, with this check, we still have the issue that https://github.com/systemd/systemd/commit/41d1f469cf10f5f3e9cb4f4853ace9b0cfe5beae was trying to solve.

> I do not think it is a good idea we assume that `$SYSTEMD_EXEC_PID=` is set for all invoked process, as it may be easily forgotten when we introduce new `execve()`

There's maybe a dozen places in the code where execve is used:
```console
$ git grep -n 'execve(at)?\([^)3]' src/
src/basic/missing_syscall.h:340:static inline int missing_execveat(int dirfd, const char *pathname,
src/core/execute.c:4610:        r = fexecve_or_execve(executable_fd, executable, final_argv, accum_env);
src/core/main.c:1538:        execve(SYSTEMD_SHUTDOWN_BINARY_PATH, (char **) command_line, env_block);
src/core/main.c:1802:                (void) execve(args[0], (char* const*) args, saved_env);
src/core/main.c:1819:                (void) execve(args[0], (char* const*) args, saved_env);
src/nspawn/nspawn.c:3457:                execve(a[0], a, env_use);
src/nspawn/nspawn.c:3460:                execve(a[0], a, env_use);
src/nspawn/nspawn.c:3463:                execve(a[0], a, env_use);
src/shared/exec-util.c:449:int fexecve_or_execve(int executable_fd, const char *executable, char *const argv[], char *const envp[]) {
src/shared/exec-util.c:451:        execveat(executable_fd, "", argv, envp, AT_EMPTY_PATH);
src/shared/exec-util.c:468:                execve(executable, argv, envp);
src/shared/exec-util.h:49:int fexecve_or_execve(int executable_fd, const char *executable, char *const argv[], char *const envp[]);
src/test/test-path-util.c:252:                r = fexecve_or_execve(fd, t, STRV_MAKE(t, "--version"), STRV_MAKE(NULL));
src/udev/udev-event.c:786:                execve(argv[0], argv, envp);
```

Reply:
OK. `get_ctty_devnr() < 0` check is dropped. And, `$SYSTEMD_EXEC_PID=` is also set for all generator invocations.